### DISCUSSION TITLE
Generic form: Why can't I use default form values of type `FieldValues`?

### DISCUSSION THREAD

**Discussion by NiklasKappel** [react-hook-form/react-hook-form#11470] [Q&A] [answered]
I am trying to write a generic `BaseForm` component, so that I can create multiple forms just by slotting in default values, validation scheme and the actual UI elements. But I did not get very far before encountering the error on line 10 in this [code sandbox](https://codesandbox.io/p/sandbox/jovial-mclaren-52gdd3?file=%2Fsrc%2FApp.tsx%3A6%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cls4ghbjw00062066k7mkvoj0%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cls4ghbjw000220667lymyesn%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cls4ghbjw00032066np9hakvi%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cls4ghbjw00052066nhmaa28d%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cls4ghbjw000220667lymyesn%2522%253A%257B%2522id%2522%253A%2522cls4ghbjw000220667lymyesn%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cls4gkq4a00022066ed6kx6a8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A6%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A6%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cls4gkq4a00022066ed6kx6a8%2522%257D%252C%2522cls4ghbjw00052066nhmaa28d%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cls4ghbjw00042066mdmfla0d%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cls4ghbjw00052066nhmaa28d%2522%252C%2522activeTabId%2522%253A%2522cls4ghbjw00042066mdmfla0d%2522%257D%252C%2522cls4ghbjw00032066np9hakvi%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cls4ghbjw00032066np9hakvi%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

The offending code is:
```TS
import { FieldValues, useForm } from "react-hook-form";

interface BaseFormProps<TFormValues extends FieldValues> {
  defaultValues: TFormValues;
}

function BaseForm<TFormValues extends FieldValues>({
  defaultValues,
}: BaseFormProps<TFormValues>) {
  const { control, handleSubmit } = useForm<TFormValues>({ defaultValues });  // ERROR
}
```

In the line marked with "ERROR", at `defaultValues`, typescript complains:
```
 Type 'TFormValues' is not assignable to type 'AsyncDefaultValues<TFormValues> | DefaultValues<TFormValues> | undefined'. typescript (2322) [10, 60]
   Type 'FieldValues' is not assignable to type 'AsyncDefaultValues<TFormValues> | DefaultValues<TFormValues> | undefined'. 

```

Note that typescript also requires `TFormValues` to extend `FieldValues`. Is this a problem with react-hook-form, or am I doing something wrong?

**Comment by ktalalykin** [comment, marked as answer]
defaultValues are not a simple object on RHF side.
if you rewrite you BaseFormProps interface like this, it should work

```
import { FieldValues, useForm, UseFormProps } from "react-hook-form";

interface BaseFormProps<TFormValues extends FieldValues> {
  defaultValues: UseFormProps<TFormValues>["defaultValues"];
}

function BaseForm<TFormValues extends FieldValues>({
  defaultValues,
}: BaseFormProps<TFormValues>) {
  const { control, handleSubmit } = useForm<TFormValues>({ defaultValues }); 
}
```
However, if you creating BaseForm to isolate hook usage, i would advise to do smth like this, it will allow to pass any attribute of `useForm` as a prop to your BaseForm

```
type BaseFormProps<TFormValues extends FieldValues> = {
  children: React.ReactNode;
} & UseFormProps<TFormValues>;

const DeclarativeFormProvider = <TFormValues extends FieldValues>(props: BaseFormProps<T>) => {
  const { children, ...formProps } = props;
  const form = useForm<T>(formProps);
 ......
};
```
